### ISSUE TITLE
Multi-Inputs CAM support

### ISSUE THREAD

**Issue by icedpanda** [frgfm/torch-cam#151]
### 🚀 Feature

Hi,  thanks for this great toolbox. 

I am wondering how to get CAM for multiple input cases. Currently, I found that the CAM I am getting is from the last input.

The model which I want to check takes 3 images as a single input and then concatenates or stack the feature maps for the final FC layer to make a multi-label classification.

Would it be possible to generate 3 CAMs for each image?

Thanks in advance 

### Motivation & pitch

- Multi-view model or multi-input model take a list of items as a single input
- Current CAM code will only generate a class activation map based on the last item in the list (depending on how items are passed to the  `forward` function tho).

**Comment by frgfm** [comment]
Hello @icedpanda 👋

Interesting use case! I'm missing some information to have a full understanding of the situation but let's try to help you 👍

I assume the input to your model is a single input tensor, either of shape:
- (N, C, H, W) where your model uses the batch dimension to stack your 3 image and generate an output of shape (K,)
- or (N, C, T, H, W) where the stack dimension is of size T, and generate an output of shape (N, K)

Now, torchcam supports models with more spatial dimension than 2 (works in 3D) for classification. As long as you can model your problem like so, it's all good. All good meaning you'll be able to get the CAM at any layer for any output class.

Back to your question, a CAM will have a shape of (N, h, w) where h and w are the spatial output size at the target layer (for 2D classification). But in the last case mentioned above, it will be (t, h, w) which would be perfect for you.

You will have to write a custom function for display though, but the cam itself will be computed by this library 🙂

Let me know if some things aren't clear!

Cheers ✌️

**Comment by icedpanda** [comment]
Hi @frgfm,

Thanks for the quick replay.

Currently, the shape of my input images is `batch x views x channels x height x width`, where `views` refers to different images.

The model would loop each view to generate embeddings and concatenates to a fully-connected layer.

Here is a sample code of my model

```python
# inputs.shape: samples x views x channels x height x width
inputs = inputs.transpose(0, 1)

return [self.feature_extractor(view_batch) for view_batch in inputs]

# than feed features to a fully-connected layer, and then the classifier layer
```
Could you please provide the sample code for displaying the heatmaps in this case?

Thanks in advance

**Comment by frgfm** [comment]
According to your snippet, I think it's better to flatten both the batch and view dimensions to get proper acceleration 👍
You'll be able to unflatten those once the model has processed them 

About your visualization, it's not exactly easy with this amount of information 😅
From your snippet the model processes your views as batches so in a completely independent way. Perhaps those are combined together later on?

**Comment by icedpanda** [comment]
I have a fully connected layer that will fuse these feature vectors and feed them to the classifier head

the architecture would looks like this

```python 

self.fusion_layer = nn.Sequential(
                nn.Linear(
                    self.num_features * self.n_views,
                    self.num_features),
                nn.ReLU(
                    inplace=True))

def forward(self, inputs):
  # inputs.shape: samples x views x channels x height x width
  inputs = inputs.transpose(0, 1)
  
  features = [self.feature_extractor(view_batch) for view_batch in inputs]
  
  # than feed features to a fully-connected layer, and then the classifier layer
  
  features = torch.cat(features, dim=1)
  
  final_features = self.fusion_layer(features)
  
  return self.classifier(final_features)
```

**Comment by frgfm** [comment]
So a few things:
- for the sake of proper interpretation of the backdrop, the first dimension of a tensor, should be the independent one (meaning that for the upcoming computation, the outputs are strictly independent with other inputs)
- keeping that in mind, you could build your features by flattening the batch and views axes before forwarding 👌
- if your condition is True, you concatenate it with channel axis. Otherwise you unflatten the first dimension into 2 axes
- now, the CAMs will be able to give you the influence on your model output at a given feature map (the cam will have N-2 dimensions where N is the number of dimension of the feature map) 
- so if you place your view dimension after the channel dimension, your problem becomes 3D, and using this library you will get a CAM of shape (V, H, W) which is what you are looking for 👍

Let me know if that helps 🙂

**Comment by frgfm** [comment]
Any update @icedpanda ? Did you manage to make it work? :)

**Comment by icedpanda** [comment]
Thanks for your help, but I haven't had time to try it yet since we are using different networks to handle different views now. So, it becomes a single input for each network that CAM already supports.
